storage/repair.go

```diff
@@ -124,23 +128,275 @@ func (r shardRepairer) Repair(
+//   - consider not always writing
+//   - factor in a minimum number of blocks repaired
+//   - and number of blocks still requiring repair
+func (r shardRepairer) repairDifferences(
 	namespace ts.ID,
 	shard databaseShard,
 	diffRes repair.MetadataComparisonResult,
+) (repair.ExecutionMetrics, error) {
+	var (
+		logger        = r.opts.InstrumentOptions().Logger()
+		session, err  = r.client.DefaultAdminSession()
+		repairSummary = repair.ExecutionMetrics{}
+		multiErr      xerrors.MultiError
+	)
+	if err != nil {
+		return repairSummary, err
+	}
+	numReplicas := session.Replicas()
+
+	// pendingBlockReplicas is a map from (series_id, block_start) -> set of host replicas for which repairs are pending
+	pendingBlockReplicas, hashToIDMap := r.newPendingReplicasMap(diffRes, session.Origin(), numReplicas)
+
+	// reqBlocks contains the same information as the above map, destructured for the `session` API
+	initialCap := diffRes.NumBlocks
+	reqBlocks := r.newRequest(pendingBlockReplicas, hashToIDMap, initialCap)
+	repairSummary.NumRequested = int64(len(reqBlocks))
+
+	// stream over the requested replicas
+	blocksIter, err := session.FetchBlocksFromPeers(namespace, shard.ID(), reqBlocks, r.rpopts.ResultOptions())
+	if err != nil {
+		return repairSummary, err
+	}
+
+	// amortizing the times marked dirty by UpdateSeries
+	dirtyBlockTimes := make(map[time.Time]struct{})
+	for blocksIter.Next() {
+		host, id, blk := blocksIter.Current()
+
+		blkID := blockID{id.Hash(), blk.StartTime()}
+		// ensure the replica was requested, should never happen
+		if hs, ok := pendingBlockReplicas[blkID]; !ok || !hs.contains(host) {
+			logger.WithFields(
+				xlog.NewLogField("id", id.String()),
+				xlog.NewLogField("host", host.String()),
+				xlog.NewLogField("blockID", blkID),
+			).Warnf("received unexpected block, session.FetchRepairBlockFromPeers violated contract, skipping.")
+			repairSummary.NumUnexpected++
+			continue
+		}
+
+		// mark the block replica as received
+		pendingBlockReplicas[blkID].remove(host)
+
+		if err := shard.UpdateSeries(id, blk); err != nil {
+			repErr := fmt.Errorf(
+				"unable to update series [ id = %v, block_start = %v, err = %v ]",
+				id.String(), blkID.start, err)
+			multiErr = multiErr.Add(repErr)
+			logger.Warnf("%v", repErr)
+			repairSummary.NumFailed++
+			continue
+		}
+
+		dirtyBlockTimes[blkID.start] = struct{}{}
+		repairSummary.NumRepaired++
+	}
+
+	// marking flush states dirty
+	dirtyBlockTimesList := make([]time.Time, 0, len(dirtyBlockTimes))
+	for t := range dirtyBlockTimes {
+		dirtyBlockTimesList = append(dirtyBlockTimesList, t)
+	}
+	shard.MarkFlushStatesDirty(dirtyBlockTimesList)
+
+	// track count for pending replicas
+	for blkID, hs := range pendingBlockReplicas {
+		if hs.empty() {
+			continue
+		}
+		for _, hbm := range hs {
+			missingErr := fmt.Errorf(
+				"replica requested but not received [ host = %v, id = %v, block_start = %v ]",
+				hbm.Host.String(), hashToIDMap[blkID.idHash], blkID.start)
```

Comment:
Probably best to `.String()` the ID here with `hashToIDMap[blkID.idHash].String()`